torch/nn/_functions/rnn.py

```diff
@@ -10,26 +10,29 @@
 
 
 def RNNReLUCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.relu(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.relu(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.tanh(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.tanh(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     if input.is_cuda:
-        igates = F.linear(input, w_ih)
+        igates = input.expand(4, input.size(0), input.size(1)).transpose(0, 1) if w_ih is None else F.linear(input,
```

Comment:
mentioned this before but using expand in the cell will break stacking with a batched input to hidden projection using skipinput, i.e. lstm_skip >>= linear >>= lstm_skip, since it's implicitly making the input matrix [I;I;I;I], which is strictly weaker than the normal projection which is untied. more concretely, with this expand skip_input implementation the linear will be nhid x nhid instead of nhid x 4*nhid.